File: app/src/main/java/org/tasks/jobs/DriveUploader.java
package org.tasks.jobs;

import static com.google.common.collect.Iterables.skip;
import static com.google.common.collect.Lists.newArrayList;
import static org.tasks.Strings.isNullOrEmpty;

import android.content.Context;
import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.work.Data;
import androidx.work.WorkerParameters;
import com.google.api.client.googleapis.json.GoogleJsonResponseException;
import com.google.api.services.drive.model.File;
import java.io.IOException;
import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import javax.inject.Inject;
import javax.net.ssl.SSLException;
import org.tasks.R;
import org.tasks.drive.DriveInvoker;
import org.tasks.injection.ApplicationContext;
import org.tasks.injection.InjectingWorker;
import org.tasks.injection.JobComponent;
import org.tasks.preferences.Preferences;
import timber.log.Timber;

public class DriveUploader extends InjectingWorker {

  private static final String FOLDER_NAME = "Tasks Backups";
  private static final String EXTRA_URI = "extra_uri";
  private static final String EXTRA_PURGE = "extra_purge";
  private static final Comparator<File> DRIVE_FILE_COMPARATOR =
      (f1, f2) -> Long.compare(f2.getModifiedTime().getValue(), f1.getModifiedTime().getValue());

  @Inject @ApplicationContext Context context;
  @Inject DriveInvoker drive;
  @Inject Preferences preferences;

  public DriveUploader(@NonNull Context context, @NonNull WorkerParameters workerParams) {
    super(context, workerParams);
  }

  static Data getInputData(Uri uri, boolean purge) {
    return new Data.Builder()
        .putString(EXTRA_URI, uri.toString())
        .putBoolean(EXTRA_PURGE, purge)
        .build();
  }

  private static List<File> getDeleteList(List<File> files) {
    Collections.sort(files, DRIVE_FILE_COMPARATOR);
    return newArrayList(skip(files, BackupWork.DAYS_TO_KEEP_BACKUP));
  }

  @Override
  protected Result run() {
    Data inputData = getInputData();
    Uri uri = Uri.parse(inputData.getString(EXTRA_URI));
    try {
      File folder = getFolder();
      preferences.setString(R.string.p_google_drive_backup_folder, folder.getId());
      drive.createFile(folder.getId(), uri);

      if (inputData.getBoolean(EXTRA_PURGE, false)) {
        List<File> files = drive.getFilesByPrefix(folder.getId(), "auto.");
        for (File file : getDeleteList(files)) {
          try {
            drive.delete(file);
          } catch (GoogleJsonResponseException e) {
            if (e.getStatusCode() == 404) {
              Timber.e(e);
            } else {
              throw e;
            }
          }
        }
      }

      return Result.success();
    } catch (SocketTimeoutException | SSLException | ConnectException | UnknownHostException e) {
      Timber.e(e);
      return Result.retry();
    } catch (IOException e) {
      firebase.reportException(e);
      return Result.failure();
    }
  }

  private File getFolder() throws IOException {
    String folderId = preferences.getStringValue(R.string.p_google_drive_backup_folder);
    File file = null;
    if (!isNullOrEmpty(folderId)) {
      try {
        file = drive.getFile(folderId);
      } catch (GoogleJsonResponseException e) {
        if (e.getStatusCode() != 404) {
          throw e;
        }
      }
    }

    return file == null || file.getTrashed() ? drive.createFolder(FOLDER_NAME) : file;
  }

  @Override
  protected void inject(JobComponent component) {
    component.inject(this);
  }
}

File: app/src/main/java/org/tasks/jobs/DriveUploader.kt
package org.tasks.jobs

import android.content.Context
import android.net.Uri
import androidx.hilt.work.HiltWorker
import androidx.work.WorkerParameters
import com.google.api.client.googleapis.json.GoogleJsonResponseException
import com.google.api.services.drive.model.File
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import org.tasks.LocalBroadcastManager
import org.tasks.R
import org.tasks.Strings.isNullOrEmpty
import org.tasks.analytics.Firebase
import org.tasks.backup.BackupConstants
import org.tasks.googleapis.InvokerFactory
import org.tasks.injection.BaseWorker
import org.tasks.preferences.Preferences
import timber.log.Timber
import java.io.FileNotFoundException
import java.io.IOException
import java.net.ConnectException
import java.net.SocketException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import javax.net.ssl.SSLException

@HiltWorker
class DriveUploader @AssistedInject constructor(
        @Assisted context: Context,
        @Assisted workerParams: WorkerParameters,
        firebase: Firebase,
        invokers: InvokerFactory,
        private val preferences: Preferences,
        private val localBroadcastManager: LocalBroadcastManager
) : BaseWorker(context, workerParams, firebase) {
    private val drive = invokers.getDriveInvoker()

    override suspend fun run(): Result {
        val inputData = inputData
        val uri = Uri.parse(inputData.getString(EXTRA_URI))
        return try {
            val folder = getFolder() ?: return Result.failure()
            preferences.setString(R.string.p_google_drive_backup_folder, folder.id)
            drive.createFile(folder.id, uri)
                    ?.let(BackupConstants::getTimestamp)
                    ?.let { preferences.setLong(R.string.p_backups_drive_last, it) }
            localBroadcastManager.broadcastPreferenceRefresh()
            if (inputData.getBoolean(EXTRA_PURGE, false)) {
                drive
                        .getFilesByPrefix(folder.id, "auto.")
                        .drop(BackupWork.DAYS_TO_KEEP_BACKUP)
                        .forEach { drive.delete(it) }
            }
            Result.success()
        } catch (e: FileNotFoundException) {
            fail(e)
        } catch (e: SocketException) {
            retry(e)
        } catch (e: SocketTimeoutException) {
            retry(e)
        } catch (e: SSLException) {
            retry(e)
        } catch (e: ConnectException) {
            retry(e)
        } catch (e: UnknownHostException) {
            retry(e)
        } catch (e: GoogleJsonResponseException) {
            when (e.statusCode) {
                401, 403 -> fail(e)
                503 -> retry(e)
                else -> retry(e, report = true)
            }
        } catch (e: IOException) {
            fail(e, report = true)
        }
    }

    private fun fail(e: Throwable, report: Boolean = false): Result {
        if (report) {
            firebase.reportException(e)
        } else {
            Timber.e(e)
        }
        return Result.failure()
    }

    private fun retry(e: Throwable, report: Boolean = false): Result {
        if (report) {
            firebase.reportException(e)
        } else {
            Timber.e(e)
        }
        return Result.retry()
    }

    @Throws(IOException::class)
    private suspend fun getFolder(): File? {
        val folderId = preferences.getStringValue(R.string.p_google_drive_backup_folder)
        var file: File? = null
        if (!isNullOrEmpty(folderId)) {
            file = drive.getFile(folderId)
        }
        return if (file == null || file.trashed) drive.createFolder(FOLDER_NAME) else file
    }

    companion object {
        private const val FOLDER_NAME = "Tasks Backups"
        const val EXTRA_URI = "extra_uri"
        const val EXTRA_PURGE = "extra_purge"
    }
}